<?xml version="1.0"?>
<!--
  Used to launch the 3dsensor launch file and rtabmap-slam
-->
<launch>
  <!-- Choose visualization -->
  <arg name="rviz" default="true" />
  
  <!-- Localization-only mode -->
  <arg   name="localization"        default="false"/>
  <arg     if="$(arg localization)" name="rtabmap_args"  default=""/>
  <arg unless="$(arg localization)" name="rtabmap_args"  default="--delete_db_on_start"/>
  
  <arg name="visual_odometry"       default="false"/>         <!-- Launch rtabmap visual odometry node -->
  <arg name="icp_odometry"          default="false"/>         <!-- Launch rtabmap icp odometry node -->
  <arg name="odom_topic"            default="odom"/>          <!-- Odometry topic name -->
  
  <group ns="rtabmap">
    <!-- SLAM (robot side) -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap_ros"    type="rtabmap" output="screen" args="$(arg rtabmap_args)">
      <param name="visual_odometry"           type="bool"   value="false"/>
      <param name="icp_odometry"              type="bool"   value="false"/>
      <param name="odom_frame_id"             type="string" value="odom"/>
      <param name="frame_id"                  type="string" value="base_footprint"/>
      <param name="wait_for_transform"        type="bool"   value="true"/>
      <param name="subscribe_depth"           type="bool"   value="true"/>
      <param name="subscribe_scan"            type="bool"   value="true"/>
      <!-- RTAB-Map's parameters: do "rosrun rtabmap rtabmap (double-dash)params" to see the list of available parameters. -->
      <param name="RGBD/ProximityBySpace"     type="string" value="false"/>
      <param name="RGBD/AngularUpdate"        type="string" value="0.01"/>
      <param name="RGBD/LinearUpdate"         type="string" value="0.01"/>
      <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>
      <param name="Reg/Force3DoF"             type="string" value="true"/>
      <param name="Vis/MinInliers"            type="string" value="12"/>
      <!-- localization mode -->
      <param     if="$(arg localization)"     name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)"     name="Mem/IncrementalMemory" type="string" value="true"/>
      <param name="Mem/InitWMWithAllNodes"                                 type="string" value="$(arg localization)"/>
      <remap from="odom"                      to="/odom"/>
      <remap from="scan"                      to="/scan"/>
      <remap from="rgb/image"                 to="/camera/rgb/image_raw"/>
      <remap from="depth/image"               to="/camera/depth/image_raw"/>
      <remap from="rgb/camera_info"           to="/camera/rgb/camera_info"/>
      <remap from="grid_map"                  to="/map"/>
      
   </node>
  </group>
  
  <!-- Visualisation RVIZ -->
  <node if="$(arg rviz)" pkg="rviz" name="rviz" type="rviz" args="-d $(find navigator_bringup)/rviz/robot_mapping.rviz" output="screen"/>
   <node pkg="nodelet" type="nodelet" name="points_xyzrgb" args="standalone rtabmap_ros/point_cloud_xyzrgb">
    <remap from="rgb/image"       to="/camera/rgb/image_raw"/>
    <remap from="depth/image"     to="/camera/depth/image_raw"/>
    <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
    
    <param name="queue_size" type="int" value="10"/>
  </node>
</launch>